text/0314-cdk-construct-hub.md

```diff
@@ -362,6 +367,96 @@ pipeline.
+we need to define a mechanism by which the front-end is able to detect and operate on various versions of them.
+
+To that end, we propose a generic mechanism to apply to all artifact types:
+
+Every artifact will comply to the [SemVer](https://semver.org/spec/v2.0.0.html) specification
+and include its major version in the `version` field of the schema:
+
+```json
+{
+  "version": "v1",
+  ...
+}
+```
+
+```json
+{
+  "version": "v2",
+  ...
+}
+```
+
+Every time a change in the artifact occurs, the contents of the artifact is replaced with the new version.
+This means that at any given point in time, the artifact can either be in its new version, or the old one.
+
+As a general guideline, we *prefer displaying an older version of an artifact, than to display nothing at all.*.
+To that end, the front-end application will have to support displaying all versions of the artifact.
+
+Specifically, we describe how to implement this for each type of artifact.
+
+##### Package Documentation
+
+Package documentation is created by the **Doc-Gen** function, which produces a JSON definition file:
+
+`packages/${assembly.name}/v${assembly.version}/docs-${lang}.json`
+
+Since the front-end fetches and parses this file, we'd like for it to operate on type safe interfaces that describe the schema.
+This will ensure that breaking changes will manifest as compile time errors, and force us to handle them.
+
+To that end, we create a dedicated package, called `construct-hub-docgen`, which will define this schema, and provide an API to produce it.
+The version of the schema file will be derived from the major version of the package.
+Every time a change is made to the schema, we automatically detect what type of change has been made. (i.e breaking or not)
+
+> This can be done by using either [json-schema-diff](https://www.npmjs.com/package/json-schema-diff), or [jsii-diff](https://github.com/aws/jsii/tree/main/packages/jsii-diff).
+
+If a breaking change is detected, we enforce that the next version to be released will include a major version bump.
+(i.e the commit was accompanied with a *BREAKING CHANGE:* notice)
+
+> This can be done by storing the version prior to the bump, and comparing with the one after.
+
+When a new major version of the schema is released, it is automatically picked up and
+deployed to the backend. Since the schema is overidden, this will break the front-end
+because the version supporting the new schema hasn't been deployed yet.
+We cannot afford this, which means we must enforce a deployment order in this scenario,
+by which the backend is deployed **only after** the front-end supporting it is deployed.
+
+We can enforce this by employing the following mechanism:
+
+- The front-end declares a `peerDepenedency` on `construct-hub-docgen`. (e.g `^1.0.0`)
+- The backend declares a `devDependency` on `construct-hub-docgen` (e.g `1.0.0`)
```

Comment:
Why pin?

```suggestion
- The backend declares a `devDependency` on `construct-hub-docgen` (e.g `^1.0.0`)
```

Reply:
I'm of the opinion that `devDependencies` should always be pinned. But we don't have to discuss this here, I think the spirit is clear, this is just an example. (both pinned and caret will work)

Reply:
I see. I guess the examples between parentheses are a little polarising/distracting. But I do get the sentiment.

Reply:
👍